src/core/seat/input-method-relay.cpp

```diff
@@ -221,6 +222,22 @@ bool wf::input_method_relay::should_grab(wlr_keyboard *kbd)
     return !is_im_sent(kbd);
 }
 
+bool wf::input_method_relay::check_superfluous_release(uint32_t key, uint32_t state)
```

Comment:
Hmm, now that you moved the `handle_key()` call to seat.cpp, doesn't the seat already do this filtering of keys for you automatically? Isn't https://github.com/WayfireWM/wayfire/pull/2111/commits/dbd98d3ae758b9a65370f760479389f24d968e41#diff-206c23fa3adf45f1a58ee9a5865dc0dec21bd2ba72a2441fb41d3fa88794cb6aR45-R59 executed for the keys here, so why need to check here again?

Note I find the whole IM stuff quite confusing so far because it seems everything is a mess (not on Wayfire's side but the protocols & clients & how they work together), so maybe my intuition is wrong ..

Also, do you have any references as to how gnome implements this or kwin? Do they also have so many checks and workarounds for the various brokenness in the protocols? I am kinda hoping there is an 'easy' solution to all the troubles that you've tried to fix here.

Reply:
> so why need to check here again?

That's exactly why again. Take the Firefox case as example:

1. No text input in focus. The user sends keydown ctrl and keydown l.
2. Firefox focuses the address as requested. Text input in focus now.
3. The user sends keyup l.
4. Input method passes through the keyup l, and sends via virtual keyboard to the compositor.
5. In the code you linked, this second keyup l is ignored because from the compositor's view it's already been released.
6. Firefox never receives the keyup l event, and so auto repeat kicks in and a lot of l is inserted.

> Also, do you have any references as to how gnome implements this or kwin?

The answer is super simple: they don't. GNOME uses D-Bus, and kwin uses input method v1 ([ref](https://fcitx-im.org/wiki/Using_Fcitx_5_on_Wayland#Support_in_Wayland_Compositor)).

Reply:
> The answer is super simple: they don't. GNOME uses D-Bus, and kwin uses input method v1 ([ref](https://fcitx-im.org/wiki/Using_Fcitx_5_on_Wayland#Support_in_Wayland_Compositor)).

Wayland at its finest :((

> 5\. In the code you linked, this second keyup l is ignored because from the compositor's view it's already been released.

Ok, so repeating the flow of events as to how I see it to avoid misunderstandings:

1. Keydown ctrl, keydown L
2. Focus text input
3. IM grabs keyboard
4. keyup L -> we go to the code I linked -> L is erased from the set -> L is sent to the IM, not to the client
5. IM sends virtual keyup L to compensate -> we go to the linked code -> L is not in the set, so it is dropped

I would say that the actual problem is that we erased L from the set in step 4. The set is supposed to contain all pressed, but not released keys for the current focus. Since the key was forwarded to the IM and not to the client, then L should have never been removed from the set.

Reply:
> Since the key was forwarded to the IM and not to the client, then L should have never been removed from the set.

Yes, it works nicely!

Reply:
Oh no. With text input focues, super+some key to switch focus will leave super pressed...

Reply:
It seems to be some bad interactions with the IM. I did some change to fcitx5 and this seemed to no longer happen.